### DISCUSSION TITLE
Optional SmallVec and SmallString support

### DISCUSSION THREAD

**Discussion by ColinTimBarndt** [diesel-rs/diesel#2893] [Ideas]
This would be a nice-to-have feature because tables with lots of small Strings or Blobs could be retrieved without allocating them on the heap. I am already using a wrapper implementation for SmallString with the features that I need (code provided below). Adding support for them behind a feature flag would be great because then no wrapper structs would be needed.

https://gist.github.com/ColinTimBarndt/28857d3f955dfd1fadea01b6fe0fe450

**Comment by weiznich** [comment]
Can you provide a benchmark showing an actual performance improvement for this feature? I'm not sure if the cost of a network call is so much larger that the allocations won't matter that much.

**Reply by ColinTimBarndt** [reply]
I could find this benchmark of Vec vs. SmallVec that showed an improvement of about 11% (704570 ns/iter vs. 627145 ns/iter). Is this what you are looking for?

> I'm not sure if the cost of a network call is so much larger that the allocations won't matter that much.

I am running a file-database with Sqlite3 on an M.2 NVMe-SSD, there are no network calls in this scenario. Though I'm not sure if this is what Diesel is used for usually.

https://github.com/servo/rust-smallvec/issues/3#issuecomment-99254932

**Reply by weiznich** [reply]
I'm well aware of this and similar benchmarks. What I'm looking for is a benchmark showing an x% improvement for a diesel loading/inserting operation, as I'm not certain that those improvements shown by the existing benchmarks won't be just noise when a database operation is involved. I'm fine if this benchmark uses a specially tuned setup like sqlite with an in memory database, but I want to see some numbers.
